### DISCUSSION TITLE
Atom: add method to run computation on inner value (atomic)?

### DISCUSSION THREAD

**Discussion by StefanBertels** [louthy/language-ext#958] [Ideas]
I use this extension method to run a computation on the `Atom` content in an atomic way:

```
public static R Eval<T, R>(this Atom<T> atom, Func<T, R> f)
{
    R result = default!;
    atom.Swap(old =>
    {
        result = f(old);
        return old;
    });
    return result;
}
```

I call this like `myAtom.Eval(_ => _.GetCurrentData())`.

My use case is a `T` which represents a ressource that isn't thread-safe (e.g. an OS handle -- might be closed on another thread => lock necessary). `R` is a value object.

Question: Is this extension a "missing" feature in LanguageExt or is it maybe  just a bad idea at all?

**Comment by louthy** [comment]
If there are side-effects within an atom's `Swap` function then all bets are off - it's very much about protecting memory (a very lightweight STM that protects a single reference), and not about protecting anything else like resource-handles.  Especially as the operation may run multiple times if there's a conflict, that could cause serious complications.  So, [this is explicitly not supported](https://github.com/louthy/language-ext/blob/main/LanguageExt.Core/Concurrency/Atom/Atom.cs#L75)

Atomic reads are always possible using `atom.Value`.  It will always give you a safe value that you can work with.  But it doesn't sound like that's what you're asking for.

**Reply by StefanBertels** [reply]
Thanks for clarification!

I was aware of the atom conflict solution behaviour "repeat on fail" and my resource had no problem with multiple execution (idempotent). But you made clear that my use case will eventually run into problems (edge cases) because of undefined order / inconsistent state (e.g. Use after Dispose, maybe even multi-threading issues). "Golden hammer issue" -- I'm using a monitor (`lock`) now.
